keps/sig-api-machinery/2339-storageversion-api-for-ha-api-servers/README.md

```diff
@@ -25,98 +50,170 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
 
 During the rolling upgrade of an HA master, the API server instances may
-use different storage versions encoding a resource. The [storageVersionHash][]
+use different storage versions encoding a resource. The [storageVersionHash]
 in the discovery document does not expose this disagreement. As a result, the
 storage migrator may proceed with migration with the false belief that all API
 server instances are encoding objects using the same storage version, resulting
-in polluted migration.  ([details][]).
+in polluted migration.  ([details]).
 
 [storageVersionHash]:https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L979
 [details]:https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/35-storage-version-hash.md#ha-masters
 
+## Motivation
+
+The lack of automated storage migration creates unfixable technical debt in the
+Kubernetes project as we are unable to drop old versions of APIs, even after the
+associated REST endpoints have been disabled.  This also blocks more nuanced
+use cases of storage migration such as encryption key rotation, which also needs
+a reliable mechanism for understanding when API servers are in agreement.
+
+### Goals
+
+- Create an API that allows an external controller to observe what storage versions are in use by API servers in an HA setup
+
+### Non-Goals
+
+- Automatically running storage migration
+
+## Proposal
+
 We propose a way to show what storage versions all API servers are using, so
 that the storage migrator can defer migration until an agreement has been
 reached.
 
-## API changes
+### Risks and Mitigations
+
+Writes to most Kubernetes resources must be prevented until the API server has
+had a chance to emit its encoding versions for all resources.  This requires us to
+have a carefully written handler that only lets a very specific set of requests through.
+Mistakes in this code could easily prevent the API server from functioning.  This
+handler will be carefully tested via unit and integration tests.
+
+## Design Details
 
 ### Resource Version API
 
 We introduce a new API `StorageVersion`, in a new API group
 `internal.apiserver.k8s.io/v1alpha1`.
 
-```golang
-//  Storage version of a specific resource.
+```go
+// Storage version of a specific resource.
 type StorageVersion struct {
-  TypeMeta
+  metav1.TypeMeta `json:",inline"`
   // The name is <group>.<resource>.
-  ObjectMeta
+  metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
-  // Spec is omitted because there is no spec field.
-  // Spec StorageVersionSpec
+  // Spec is an empty spec. It is here to comply with Kubernetes API style.
+  Spec StorageVersionSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
```

Comment:
I don't see why we would need this, but we settle these things in API review, not KEPs

Reply:
@lavalamp this is copy pasted from the current code in k/k

Reply:
haha ok